### ISSUE TITLE
[bug/fix] Ray Serve container should be able to see and use the GPU (via PyTorch CUDA).

### ISSUE THREAD

**Issue by daeun-ops** [daeun-ops/hybrid-mlops-demo#16]
## Goal / Why
Right now, `torch.cuda.is_available()` is returning `False` inside the container even though the host has a compatible driver/runtime stack:

* NVIDIA Driver: 512.89
* CUDA Version (Driver): 11.6
* Container Base: nvidia/cuda:11.6.2-base-ubuntu20.04
* PyTorch: 1.13.1+cu116

This stack is theoretically compatible, so GPU should be visible. Lack of GPU visibility means:

* The model currently runs on CPU only
* We can't measure real GPU inference latency / utilization
* Airflow DAG validation of "train → deploy → infer on GPU" is blocked

We want full GPU access so we can demonstrate real production-style GPU inference with Ray Serve.

## Scope

What needs to change:

* Docker runtime config:

  * Ensure the container is started with GPU access (e.g. `--gpus all` or `deploy.resources.reservations.devices` in compose)
  * Make sure `nvidia-container-toolkit` / `nvidia-container-runtime` is actually being used, not plain `runc`
* docker-compose.yml:

  * Add proper `runtime` / `device` / `env` entries so Ray Serve can see `/dev/nvidia*`
* Image:

  * Confirm the image still uses CUDA 11.6 base and installs `torch==1.13.1+cu116`, not CPU-only torch
  * Confirm `pynvml` (optional) for metrics export
* Service touched:

  * `ray-inference` (Ray Serve container)
  * compose / runtime for local dev
  * Airflow DAG that calls this service for inference validation

No other services (MLflow, Airflow scheduler/web, etc.) should need changes right now.

## How to Test

1. Build and run locally with docker compose:

   * `docker compose build ray-inference`
   * `docker compose up -d ray-inference`
     (container MUST be started with GPU access)

2. Exec into the container and verify CUDA from PyTorch:

   ```bash
   docker exec -it ray-inference python3 - <<'PY'
   import torch
   print("cuda_available:", torch.cuda.is_available())
   print("device_count:", torch.cuda.device_count())
   if torch.cuda.is_available():
       print("device_name:", torch.cuda.get_device_name(0))
   PY
   ```

3. Still in the container, confirm the driver is exposed:

   ```bash
   docker exec -it ray-inference bash -lc 'nvidia-smi || echo "nvidia-smi not found in image"'
   ```

4. Call the model endpoint and confirm it actually runs on GPU:

   * `curl http://localhost:8000/infer` (or whatever the current Ray Serve route is)
   * Check logs from `ray-inference` container for `cuda:` device usage

5. (Optional) Trigger the Airflow DAG that does:

   * training step
   * inference request to Ray Serve
   * write result + timing to MLflow
     Expectation: the inference task should report GPU execution, not CPU fallback.

## Risk / Notes

* Assumes host already has working NVIDIA driver + WSL/WSL2 passthrough (Windows 11 + WSL2 case).
* Requires `nvidia-container-toolkit` to be installed and actually wired into Docker Desktop / Docker Engine. If that's missing, container will start but with zero `/dev/nvidia*` devices.
* Local dev only right now. No production cluster scheduling yet (no K8s `nvidia.com/gpu` resources involved).
* If GPU is not available, Ray Serve should still boot on CPU — so rollback is basically "run without GPU flags."

Secrets / credentials:

* None. This is pure infra/runtime, not credentials.

Impact on other services:

* MLflow, Airflow, etc. only read metrics/logs, so low blast radius.

## Milestone

Inference Observability v0.1
This is part of proving end-to-end:

* train → deploy (Ray Serve) → infer on GPU
* collect latency / utilization metrics
* expose this pipeline for portfolio/demo

**Comment by daeun-ops** [comment]
<img width="444" height="175" alt="Image" src="https://github.com/user-attachments/assets/d30136bd-98fb-4f4b-b2c9-b67eb1e86fbc" />

**Comment by daeun-ops** [comment]
<img width="513" height="237" alt="Image" src="https://github.com/user-attachments/assets/50d8c952-858b-4fff-9a37-ecfbfa09e9bb" />

**Comment by daeun-ops** [comment]
<img width="949" height="330" alt="Image" src="https://github.com/user-attachments/assets/a648798c-7e61-4219-9df3-f42cceff36f9" />
로그가 바뀌는건 고마운데 점점 내가 뭘 해결하고 있는지 모르겠어

**Comment by daeun-ops** [comment]
<img width="765" height="69" alt="Image" src="https://github.com/user-attachments/assets/b04ff4e1-b2ef-42f3-a9f4-3ad97a674ca8" />

<img width="957" height="76" alt="Image" src="https://github.com/user-attachments/assets/4daebaa4-db8d-433f-9051-c703f052f9b8" />

**Comment by daeun-ops** [comment]
<img width="765" height="202" alt="Image" src="https://github.com/user-attachments/assets/dfb157eb-a3a2-4cea-a805-fa1280ccd7f0" />

컨테이너 내부에서도 확인 완료 

<img width="806" height="360" alt="Image" src="https://github.com/user-attachments/assets/2684cb5f-3c21-43f3-8040-a264b8056d88" />
